test/utils/Decimal.ts

```diff
@@ -1,7 +1,201 @@
+    }
+    return one;
+  }
+
+  /** @return decimal(this) + decimal(x) */
+  public add(x:Numberish): Decimal {
+    return this.toDecimal( this.int + this.toDecimalBN(x) );
+  }
+
+  /** @return decimal(this) - decimal(x) */
+  public sub(x:Numberish): Decimal {
+    return this.toDecimal( this.int - this.toDecimalBN(x) );
+  }
+
+  /** @return decimal(this) * decimal(x) */
+  public mul(x:Numberish): Decimal {
+    // mulf = (a * b) / ONE
+    return this.toDecimal( (this.int * this.toDecimalBN(x)) / this.one() );
+  }
+
+  /** @return decimal(this) / decimal(x) */
+  public div(x:Numberish): Decimal {
+    // divf = (a * ONE) / b
+    return this.toDecimal( (this.int * this.one()) / this.toDecimalBN(x) );
+  }
+
+  private static toFixedPointInteger(value:Numberish, decimals:number): bigint {
+    if (typeof(value) === "bigint") {
+      return value; // accept BigInt without any validation, this is necessary to enable raw interop
+    }
+
+    if (value instanceof BigNumber) {
+      return BigInt(value.toString()); // accept BigNumber without any validation
+    }
+
+    // for Decimal types we can perform optimized upscaling/downscaling fastpaths
+    if (value instanceof Decimal) {
+      if (value.decimals === decimals) {
+        return value.int; // this is a no-op case
+      } else if (value.decimals > decimals) {
+        // incoming needs to be truncated
+        const downscaleDigits = value.decimals - decimals;
+        return value.int / this._one(downscaleDigits);
+      } else {
+        // incoming needs to be upscaled
+        const upscaleDigits = decimals - value.decimals;
+        return value.int * this._one(upscaleDigits);
+      }
+    }
+
+    // get the string representation of the Numberish value
+    const val = value.toString();
+    // figure out if there is a fractional part to it and get the Whole part
+    const decimalIdx = val.indexOf('.');
+    const whole = val.slice(0, decimalIdx === -1 ? val.length : decimalIdx);
+
+    if (decimals === 0) { // pure integer case, TRUNCATE any decimals
+      return BigInt(whole);
+    }
+
+    if (decimalIdx === -1) { // input was integer eg "1234"
+      return BigInt(val.padEnd(val.length + decimals, "0"));
+    }
+
+    // input was a decimal eg "123.45678"
+    // extract the fractional part of the decimal string
+    const fract = val.slice(decimalIdx+1, Math.min(decimalIdx+1+decimals, val.length));
+    // if it's not long enough, create a trail of zeroes to satisfy decimals precision
+    const trail = decimals > fract.length ? decimals - fract.length : 0;
+    return BigInt(whole + fract.padEnd(fract.length + trail, "0"));
+  }
+
+  private _toString(maxDecimals:number, round:boolean=false): string {
+    if (this.decimals === 0) {
+      return this.int.toString();
+    }
+
+    // get the BN digits and check if it's negative
+    const s = this.int.toString();
+    const neg = s[0] === '-';
+    const abs = neg ? s.slice(1) : s;
```

Comment:
Sup? Why is all these done on strings?

```suggestion
    const neg = this.int < 0n;
    const abs = neg ? (this.int * -1n) : this.int;
```

Reply:
Mostly because the rest of the method operates on the `abs: string`. It would have to look like this:
```ts
    const neg = this.int < 0;
    const abs = (neg ? this.int * BigInt(-1) : this.int).toString();
```
Probably this.int * BigInt(-1) will be faster because s.slice(1) makes a new substring